packages/core/src/extensions/ImageToolbar/ImageToolbarPlugin.ts

```diff
@@ -0,0 +1,247 @@
+import { Node as PMNode } from "prosemirror-model";
+import { EditorState, Plugin, PluginKey } from "prosemirror-state";
+import { EditorView } from "prosemirror-view";
+import {
+  BaseUiElementCallbacks,
+  BaseUiElementState,
+  BlockNoteEditor,
+  BlockSchema,
+  BlockSpec,
+  SpecificBlock,
+} from "../..";
+import { EventEmitter } from "../../shared/EventEmitter";
+
+export type ImageToolbarCallbacks = BaseUiElementCallbacks;
+
+export type ImageToolbarState = BaseUiElementState & {
+  block: SpecificBlock<
+    BlockSchema & {
+      image: BlockSpec<
+        "image",
+        {
+          src: { default: string };
+        },
+        false
+      >;
+    },
+    "image"
+  >;
+};
+
+export class ImageToolbarView {
+  private imageToolbarState?: ImageToolbarState;
+  public updateImageToolbar: () => void;
+
+  public prevWasEditable: boolean | null = null;
+
+  public shouldShow: (state: EditorState) => boolean = (state) =>
+    "node" in state.selection &&
+    (state.selection.node as PMNode).type.name === "image" &&
+    (state.selection.node as PMNode).attrs.src === "";
+
+  constructor(
+    // private readonly editor: BlockNoteEditor<BSchema>,
+    private readonly pluginKey: PluginKey,
+    private readonly pmView: EditorView,
+    updateImageToolbar: (imageToolbarState: ImageToolbarState) => void
+  ) {
+    this.updateImageToolbar = () => {
+      if (!this.imageToolbarState) {
+        throw new Error("Attempting to update uninitialized image toolbar");
+      }
+
+      updateImageToolbar(this.imageToolbarState);
+    };
+
+    pmView.dom.addEventListener("mousedown", this.mouseDownHandler);
+
+    pmView.dom.addEventListener("dragstart", this.dragstartHandler);
+
+    pmView.dom.addEventListener("blur", this.blurHandler);
+
+    document.addEventListener("scroll", this.scrollHandler);
+  }
+
+  mouseDownHandler = () => {
+    if (this.imageToolbarState?.show) {
+      this.imageToolbarState.show = false;
+      this.updateImageToolbar();
+    }
+  };
+
+  // For dragging the whole editor.
+  dragstartHandler = () => {
+    if (this.imageToolbarState?.show) {
+      this.imageToolbarState.show = false;
+      this.updateImageToolbar();
+    }
+  };
+
+  blurHandler = (event: FocusEvent) => {
+    const editorWrapper = this.pmView.dom.parentElement!;
+
+    // Checks if the focus is moving to an element outside the editor. If it is,
+    // the toolbar is hidden.
+    if (
+      // An element is clicked.
+      event &&
+      event.relatedTarget &&
+      // Element is inside the editor.
+      (editorWrapper === (event.relatedTarget as Node) ||
+        editorWrapper.contains(event.relatedTarget as Node))
+    ) {
+      return;
+    }
+
+    if (this.imageToolbarState?.show) {
+      this.imageToolbarState.show = false;
+      this.updateImageToolbar();
+    }
+  };
+
+  scrollHandler = () => {
+    if (this.imageToolbarState?.show) {
+      const blockElement = document.querySelector(
```

Comment:
It's only different because we need to get a different bounding box depending on the UI element opened:
- Formatting Toolbar: Editor selection bounding box
- Hyperlink Toolbar: Hyperlink mark bounding box
- Side Menu: Hovered block content bounding box
- Slash menu: Query start selection bounding box
- Image Toolbar: Bounding box of the image block that triggered its opening

Pretty much each of them have a different way of finding the correct bounding box. For the image toolbar it just made the most sense to do it by providing the block in the transaction that triggers it to open. Then the plugin can find the element in the DOM by the block ID and get the bounding box.

Reply:
Exactly, the side menu can easily figure out both when to show/hide itself as well as which block is "selected" on its own. The image toolbar would have a much harder job of that (listen for all clicks in the editor, get the clicked image block, has to check if the image has no `src` - also a completely different process also when inserting an image using the slash menu), so this solution is a lot more simple and elegant.